File: src/approvals.js
/**
 * Durable HITL wait snapshot — idle-resume pattern (Stately/Eve inspired).
 * Also holds physical claim/complete so body work closes like auth.
 * Pure state transitions; IO lives in turn.js / CLI.
 */

/** @typedef {'pending' | 'approved' | 'denied'} ApprovalStatus */
/** @typedef {'open' | 'claimed' | 'completed'} PhysicalStatus */
/** @typedef {'idle_waiting' | 'clear' | 'partial'} SnapshotStatus */

/**
 * @returns {object} empty snapshot
 */
export function emptySnapshot(opts = {}) {
  return {
    version: 1,
    status: 'clear',
    phase: 'IDLE',
    pending: [],
    physical: [],
    history: [],
    updatedAt: opts.now || new Date().toISOString(),
    meta: { source: opts.source || 'ensembly' },
  };
}

/**
 * Build or merge pending authorizations from HITL-classified actions.
 * @param {Array<object>} actions enriched with classification
 * @param {object} [existing]
 */
export function upsertPendingFromActions(actions = [], existing = null, opts = {}) {
  const snap = existing ? structuredClone(existing) : emptySnapshot(opts);
  const now = opts.now || new Date().toISOString();
  const byId = new Map((snap.pending || []).map((p) => [p.id, p]));

  for (const action of actions) {
    const needs =
      action.classification?.hitl === true ||
      action.hitl === true ||
      Boolean(action.kind && needsApprovalKind(action.kind));
    if (!needs) continue;

    const id = `auth-${action.id || action.title || Math.random().toString(36).slice(2, 8)}`;
    const key = action.id ? `auth-${action.id}` : id;
    const prev = byId.get(key);
    if (prev && (prev.status === 'approved' || prev.status === 'denied')) {
      // Keep resolved; do not re-open automatically
      continue;
    }
    byId.set(key, {
      id: key,
      actionId: action.id || null,
      title: action.title || action.id || key,
      kind: action.kind || 'hitl',
      area: action.area || null,
      realm: action.realm || action.realmInfo?.realm || null,
      status: prev?.status || 'pending',
      createdAt: prev?.createdAt || now,
      updatedAt: now,
      reason: action.classification?.reason || 'HITL required',
    });
  }

  snap.pending = [...byId.values()];
  snap.updatedAt = now;
  snap.status = deriveStatus(snap.pending);
  snap.phase = snap.status === 'idle_waiting' ? 'HITL_WAIT' : snap.status === 'clear' ? 'CLEAR' : 'PARTIAL';
  return snap;
}

function needsApprovalKind(kind) {
  return [
    'external_email_send',
    'job_application_submit',
    'calendar_mutate',
    'finance_transfer',
    'git_push_shared',
    'publish_private_data',
  ].includes(kind);
}

/**
 * @param {Array<{ status: string }>} pending
 */
export function deriveStatus(pending = []) {
  const open = pending.filter((p) => p.status === 'pending');
  if (open.length === 0) return 'clear';
  const resolved = pending.filter((p) => p.status === 'approved' || p.status === 'denied');
  if (resolved.length > 0 && open.length > 0) return 'partial';
  return 'idle_waiting';
}

/**
 * Apply approve or deny — advances wait snapshot.
 * @param {object} snapshot
 * @param {string} approvalId
 * @param {'approve' | 'deny'} decision
 */
export function applyDecision(snapshot, approvalId, decision, opts = {}) {
  if (!snapshot || typeof snapshot !== 'object') {
    throw new Error('snapshot required');
  }
  if (decision !== 'approve' && decision !== 'deny') {
    throw new Error('decision must be approve or deny');
  }
  const now = opts.now || new Date().toISOString();
  const next = structuredClone(snapshot);
  const item = (next.pending || []).find((p) => p.id === approvalId || p.actionId === approvalId);
  if (!item) {
    throw new Error(`approval not found: ${approvalId}`);
  }
  if (item.status !== 'pending') {
    throw new Error(`approval already ${item.status}: ${item.id}`);
  }
  item.status = decision === 'approve' ? 'approved' : 'denied';
  item.updatedAt = now;
  item.decidedBy = opts.actor || 'operator';
  next.history = [
    ...(next.history || []),
    { id: item.id, decision: item.status, at: now, actor: item.decidedBy },
  ];
  next.updatedAt = now;
  next.status = deriveStatus(next.pending);
  next.phase = next.status === 'idle_waiting' ? 'HITL_WAIT' : next.status === 'clear' ? 'CLEAR' : 'PARTIAL';
  return next;
}

/**
 * Pending authorizations only.
 */
export function listPending(snapshot) {
  return (snapshot?.pending || []).filter((p) => p.status === 'pending');
}

/**
 * Upsert physical rows from realm-physical actions (do not reopen completed).
 * @param {Array<object>} actions
 * @param {object|null} existing
 */
export function upsertPhysicalFromActions(actions = [], existing = null, opts = {}) {
  const snap = existing ? structuredClone(existing) : emptySnapshot(opts);
  if (!Array.isArray(snap.physical)) snap.physical = [];
  const now = opts.now || new Date().toISOString();
  const byId = new Map(snap.physical.map((p) => [p.id, p]));

  for (const action of actions) {
    const realm = action.realm || action.realmInfo?.realm;
    if (realm !== 'physical') continue;
    const id = action.id || action.title;
    if (!id) continue;
    const prev = byId.get(id);
    if (prev && prev.status === 'completed') {
      // Keep closed; do not re-open automatically
      continue;
    }
    byId.set(id, {
      id,
      title: action.title || id,
      area: action.area || null,
      status: prev?.status || 'open',
      createdAt: prev?.createdAt || now,
      updatedAt: now,
    });
  }

  snap.physical = [...byId.values()];
  snap.updatedAt = now;
  return snap;
}

/**
 * Apply claim | complete | release on a physical action id.
 * @param {object} snapshot
 * @param {string} actionId
 * @param {'claim' | 'complete' | 'release'} decision
 */
export function applyPhysicalDecision(snapshot, actionId, decision, opts = {}) {
  if (!snapshot || typeof snapshot !== 'object') {
    throw new Error('snapshot required');
  }
  if (!['claim', 'complete', 'release'].includes(decision)) {
    throw new Error('decision must be claim, complete, or release');
  }
  const now = opts.now || new Date().toISOString();
  const next = structuredClone(snapshot);
  if (!Array.isArray(next.physical)) next.physical = [];

  let item = next.physical.find((p) => p.id === actionId);
  if (!item) {
    // Allow claim/complete of an id not yet upserted (operator names the act)
    item = {
      id: actionId,
      title: opts.title || actionId,
      area: opts.area || null,
      status: 'open',
      createdAt: now,
      updatedAt: now,
    };
    next.physical.push(item);
  }

  if (decision === 'claim') {
    if (item.status === 'completed') {
      throw new Error(`physical already completed: ${item.id}`);
    }
    item.status = 'claimed';
  } else if (decision === 'complete') {
    item.status = 'completed';
  } else {
    // release → open again (unless we want complete sticky — release only from claimed)
    if (item.status === 'completed') {
      throw new Error(`cannot release completed physical: ${item.id}`);
    }
    item.status = 'open';
  }
  item.updatedAt = now;
  item.actor = opts.actor || 'operator';
  next.history = [
    ...(next.history || []),
    { id: item.id, decision: `physical_${decision}`, at: now, actor: item.actor },
  ];
  next.updatedAt = now;
  return next;
}

/**
 * Physical rows that are still open or claimed (not completed).
 */
export function listActivePhysical(snapshot) {
  return (snapshot?.physical || []).filter((p) => p.status === 'open' || p.status === 'claimed');
}

/**
 * Map of physical id → status from snapshot.
 */
export function physicalStatusMap(snapshot) {
  const map = new Map();
  for (const p of snapshot?.physical || []) {
    map.set(p.id, p.status);
  }
  return map;
}

/**
 * JSON-serializable check (round-trip).
 */
export function serializeSnapshot(snapshot) {
  return JSON.stringify(snapshot);
}

export function parseSnapshot(json) {
  const obj = typeof json === 'string' ? JSON.parse(json) : json;
  if (!obj || obj.version !== 1) {
    throw new Error('unsupported snapshot version');
  }
  // Backward compatible: older snapshots lack physical[]
  if (!Array.isArray(obj.physical)) obj.physical = [];
  return obj;
}

File: crates/peram-kernel/src/approvals.rs
//! Durable HITL wait snapshot + physical claim lifecycle.
//! Port of ensembly `src/approvals.js` core transitions.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ApprovalStatus {
    Pending,
    Approved,
    Denied,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PhysicalStatus {
    Open,
    Claimed,
    Completed,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum SnapshotStatus {
    IdleWaiting,
    Clear,
    Partial,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PendingAuth {
    pub id: String,
    pub action_id: Option<String>,
    pub title: String,
    pub kind: String,
    pub area: Option<String>,
    pub realm: Option<String>,
    pub status: ApprovalStatus,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub reason: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub place: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub digital_flow_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub flow_status: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PhysicalRow {
    pub id: String,
    pub title: String,
    pub area: Option<String>,
    pub status: PhysicalStatus,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HistoryEntry {
    pub id: String,
    pub decision: String,
    pub at: DateTime<Utc>,
    pub actor: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Snapshot {
    pub version: u32,
    pub status: SnapshotStatus,
    pub phase: String,
    pub pending: Vec<PendingAuth>,
    pub physical: Vec<PhysicalRow>,
    pub history: Vec<HistoryEntry>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Error)]
pub enum ApprovalError {
    #[error("approval not found: {0}")]
    NotFound(String),
    #[error("approval already {0}")]
    AlreadyResolved(String),
    #[error("invalid decision: {0}")]
    InvalidDecision(String),
}

impl Snapshot {
    pub fn empty(now: DateTime<Utc>) -> Self {
        Self {
            version: 1,
            status: SnapshotStatus::Clear,
            phase: "CLEAR".into(),
            pending: vec![],
            physical: vec![],
            history: vec![],
            updated_at: now,
        }
    }
}

pub fn derive_status(pending: &[PendingAuth]) -> SnapshotStatus {
    let open: Vec<_> = pending
        .iter()
        .filter(|p| p.status == ApprovalStatus::Pending)
        .collect();
    if open.is_empty() {
        return SnapshotStatus::Clear;
    }
    let resolved = pending.iter().any(|p| {
        matches!(
            p.status,
            ApprovalStatus::Approved | ApprovalStatus::Denied
        )
    });
    if resolved {
        SnapshotStatus::Partial
    } else {
        SnapshotStatus::IdleWaiting
    }
}

fn phase_for(status: &SnapshotStatus) -> &'static str {
    match status {
        SnapshotStatus::IdleWaiting => "HITL_WAIT",
        SnapshotStatus::Clear => "CLEAR",
        SnapshotStatus::Partial => "PARTIAL",
    }
}

pub fn list_pending(snap: &Snapshot) -> Vec<&PendingAuth> {
    snap.pending
        .iter()
        .filter(|p| p.status == ApprovalStatus::Pending)
        .collect()
}

/// Upsert pending from HITL actions.
pub fn upsert_pending_from_actions(
    actions: &[(String, String, String, bool)], // id, title, kind, hitl
    existing: Option<Snapshot>,
    now: DateTime<Utc>,
) -> Snapshot {
    let mut snap = existing.unwrap_or_else(|| Snapshot::empty(now));
    for (id, title, kind, hitl) in actions {
        if !hitl {
            continue;
        }
        let key = format!("auth-{id}");
        if let Some(prev) = snap.pending.iter().find(|p| p.id == key) {
            if matches!(
                prev.status,
                ApprovalStatus::Approved | ApprovalStatus::Denied
            ) {
                continue;
            }
        }
        let created = snap
            .pending
            .iter()
            .find(|p| p.id == key)
            .map(|p| p.created_at)
            .unwrap_or(now);
        snap.pending.retain(|p| p.id != key);
        snap.pending.push(PendingAuth {
            id: key,
            action_id: Some(id.clone()),
            title: title.clone(),
            kind: kind.clone(),
            area: None,
            realm: None,
            status: ApprovalStatus::Pending,
            created_at: created,
            updated_at: now,
            reason: "HITL required".into(),
            place: None,
            digital_flow_id: None,
            flow_status: None,
        });
    }
    snap.status = derive_status(&snap.pending);
    snap.phase = phase_for(&snap.status).into();
    snap.updated_at = now;
    snap
}

pub fn apply_decision(
    snap: &Snapshot,
    approval_id: &str,
    decision: &str,
    actor: &str,
    now: DateTime<Utc>,
) -> Result<Snapshot, ApprovalError> {
    let mut next = snap.clone();
    let item = next
        .pending
        .iter_mut()
        .find(|p| p.id == approval_id || p.action_id.as_deref() == Some(approval_id))
        .ok_or_else(|| ApprovalError::NotFound(approval_id.into()))?;
    if item.status != ApprovalStatus::Pending {
        return Err(ApprovalError::AlreadyResolved(format!("{:?}", item.status)));
    }
    item.status = match decision {
        "approve" => ApprovalStatus::Approved,
        "deny" => ApprovalStatus::Denied,
        other => return Err(ApprovalError::InvalidDecision(other.into())),
    };
    item.updated_at = now;
    let status_str = match item.status {
        ApprovalStatus::Approved => "approved",
        ApprovalStatus::Denied => "denied",
        ApprovalStatus::Pending => "pending",
    };
    let id = item.id.clone();
    next.history.push(HistoryEntry {
        id,
        decision: status_str.into(),
        at: now,
        actor: actor.into(),
    });
    next.status = derive_status(&next.pending);
    next.phase = phase_for(&next.status).into();
    next.updated_at = now;
    Ok(next)
}

pub fn apply_physical_decision(
    snap: &Snapshot,
    action_id: &str,
    decision: &str,
    now: DateTime<Utc>,
) -> Result<Snapshot, ApprovalError> {
    let mut next = snap.clone();
    let idx = next.physical.iter().position(|p| p.id == action_id);
    if idx.is_none() {
        next.physical.push(PhysicalRow {
            id: action_id.into(),
            title: action_id.into(),
            area: None,
            status: PhysicalStatus::Open,
            created_at: now,
            updated_at: now,
        });
    }
    let item = next
        .physical
        .iter_mut()
        .find(|p| p.id == action_id)
        .unwrap();
    match decision {
        "claim" => item.status = PhysicalStatus::Claimed,
        "complete" => item.status = PhysicalStatus::Completed,
        "release" => item.status = PhysicalStatus::Open,
        other => return Err(ApprovalError::InvalidDecision(other.into())),
    }
    item.updated_at = now;
    next.updated_at = now;
    Ok(next)
}

pub fn upsert_physical(
    actions: &[(String, String)], // id, title physical only
    existing: Option<Snapshot>,
    now: DateTime<Utc>,
) -> Snapshot {
    let mut snap = existing.unwrap_or_else(|| Snapshot::empty(now));
    for (id, title) in actions {
        if let Some(prev) = snap.physical.iter().find(|p| p.id == *id) {
            if prev.status == PhysicalStatus::Completed {
                continue;
            }
        }
        let created = snap
            .physical
            .iter()
            .find(|p| p.id == *id)
            .map(|p| p.created_at)
            .unwrap_or(now);
        let status = snap
            .physical
            .iter()
            .find(|p| p.id == *id)
            .map(|p| p.status.clone())
            .unwrap_or(PhysicalStatus::Open);
        snap.physical.retain(|p| p.id != *id);
        snap.physical.push(PhysicalRow {
            id: id.clone(),
            title: title.clone(),
            area: None,
            status,
            created_at: created,
            updated_at: now,
        });
    }
    snap.updated_at = now;
    snap
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::TimeZone;

    fn now() -> DateTime<Utc> {
        Utc.with_ymd_and_hms(2026, 7, 15, 12, 0, 0).unwrap()
    }

    #[test]
    fn approve_closes_pending() {
        let snap = upsert_pending_from_actions(
            &[("apply".into(), "Apply".into(), "job_application_submit".into(), true)],
            None,
            now(),
        );
        assert_eq!(list_pending(&snap).len(), 1);
        let next = apply_decision(&snap, "auth-apply", "approve", "operator", now()).unwrap();
        assert!(list_pending(&next).is_empty());
        assert_eq!(next.pending[0].status, ApprovalStatus::Approved);
    }

    #[test]
    fn claim_complete_physical() {
        let snap = upsert_physical(&[("grocery".into(), "Grocery".into())], None, now());
        let c = apply_physical_decision(&snap, "grocery", "claim", now()).unwrap();
        assert_eq!(c.physical[0].status, PhysicalStatus::Claimed);
        let d = apply_physical_decision(&c, "grocery", "complete", now()).unwrap();
        assert_eq!(d.physical[0].status, PhysicalStatus::Completed);
    }
}
